### ISSUE TITLE
vfs server does not sufficiently enforce cap session quotas

### ISSUE THREAD

**Issue by ehmry** [genodelabs/genode#2553]
~~The VFS server tracks the RAM usage of each client and forwards OOM exceptions to clients. In theory this allows the server to consume arbitrary amounts of RAM on behalf of a client, so long as the client will transfer additional RAM quota when indicated by an OOM failure from the server. However the same is not implemented for capability quotas, and each RAM dataspace allocation consumes at least one capability. So when the server runs out of capabilities by action of one client, the server will block every client as it waits for a quota upgrade from the parent.~~

The VFS server does enforce the upfront capability cost of sessions against session quotas, nor does it enforce sessions capability quotas during session lifetimes.

**Comment by ehmry** [comment]
Fixed at 132f2838261052bb34e64378661032f477c65b38 and 25af5e5fa78c3483dbb24a94656bf819ba73a0d2. The follow-up task is to check the vfs plugins for any persistent allocations made against session quotas.

**Comment by ehmry** [comment]
Fix without operator commit at 19bbaef5d29fc0c2bc8eeb157c80a0c08df7b058.

**Comment by ehmry** [comment]
Rebased after #2544 with naming fixup at f37852b285557df5c04a31e143a22c4256402f0d.

**Comment by ehmry** [comment]
@chelmuth pointed out to me that directory reads are now missing, and I've discovered that this issue is the cause. The Dir_file_system plugin allocates directory handles during directory reads, and should an allocation fail during read-by-packet, the an opaque error is returned to the application and the directory read is truncated.
